File: src/D3D11/D3D11Pipeline.cpp
#include "D3D11Pipeline.h"
#include "D3D11ShaderModule.h"
#include "D3D11Utils.h"

namespace grabs::D3D11
{
    D3D11Pipeline::D3D11Pipeline(ID3D11Device* device, const PipelineDescription& description)
    {
        auto vertexModule = dynamic_cast<D3D11ShaderModule*>(description.VertexShader);
        auto vertData = vertexModule->Blob->GetBufferPointer();
        auto vertSize = vertexModule->Blob->GetBufferSize();

        D3D11_CHECK_RESULT(device->CreateVertexShader(vertData, vertSize,nullptr, &VertexShader));

        auto pixelModule = dynamic_cast<D3D11ShaderModule*>(description.PixelShader);
        auto pixlData = pixelModule->Blob->GetBufferPointer();
        auto pixlSize = pixelModule->Blob->GetBufferSize();

        D3D11_CHECK_RESULT(device->CreatePixelShader(pixlData, pixlSize,nullptr, &PixelShader));

        std::vector<D3D11_INPUT_ELEMENT_DESC> elementDescs;
        elementDescs.reserve(description.InputLayout.size());

        uint32_t i = 0;
        for (const auto& layout : description.InputLayout)
        {
            D3D11_INPUT_ELEMENT_DESC desc
            {
                .SemanticName = "TEXCOORD",
                .SemanticIndex = i++,
                .Format = Utils::FormatToD3D(layout.Format),
                .InputSlot = layout.Slot,
                .AlignedByteOffset = layout.Offset,
                .InputSlotClass = layout.Type == InputType::PerInstance ? D3D11_INPUT_PER_INSTANCE_DATA : D3D11_INPUT_PER_VERTEX_DATA,
                .InstanceDataStepRate = 0
            };

            elementDescs.push_back(desc);
        }

        D3D11_CHECK_RESULT(device->CreateInputLayout(elementDescs.data(), elementDescs.size(), vertData, vertSize, &InputLayout));
    }

    D3D11Pipeline::~D3D11Pipeline()
    {
        InputLayout->Release();
        PixelShader->Release();
        VertexShader->Release();
    }
}

File: src/grabs.Graphics/D3D11/D3D11Pipeline.cs
using Vortice.Direct3D11;

namespace grabs.Graphics.D3D11;

internal sealed class D3D11Pipeline : Pipeline
{
    public readonly ID3D11VertexShader VertexShader;

    public readonly ID3D11PixelShader PixelShader;

    public readonly ID3D11InputLayout? Layout;

    public readonly Dictionary<uint, VertexBufferInfo>? VertexBuffers;
    
    public D3D11Pipeline(ID3D11Device device, ref readonly PipelineInfo info)
    {
        D3D11ShaderModule vertexModule = (D3D11ShaderModule) info.VertexShader;
        D3D11ShaderModule pixelModule = (D3D11ShaderModule) info.PixelShader;

        VertexShader = device.CreateVertexShader(vertexModule.Blob);
        PixelShader = device.CreatePixelShader(pixelModule.Blob);

        if (info.VertexBuffers.Length > 0)
        {
            VertexBuffers = new Dictionary<uint, VertexBufferInfo>(info.VertexBuffers.Length);

            foreach (VertexBufferInfo bufferInfo in info.VertexBuffers)
                VertexBuffers.Add(bufferInfo.Binding, bufferInfo);
        }

        if (info.InputLayout.Length > 0)
        {
            InputElementDescription[] elementDescs = new InputElementDescription[info.InputLayout.Length];

            for (int i = 0; i < info.InputLayout.Length; i++)
            {
                ref readonly InputLayoutInfo layout = ref info.InputLayout[i];

                elementDescs[i] = new InputElementDescription()
                {
                    SemanticName = "TEXCOORD",
                    SemanticIndex = (uint) i,
                    Format = layout.Format.ToD3D(),
                    Slot = layout.Slot,
                    AlignedByteOffset = layout.Offset,
                    Classification = InputClassification.PerVertexData
                };
            }

            Layout = device.CreateInputLayout(elementDescs, vertexModule.Blob);
        }
    }
    
    public override void Dispose()
    {
        Layout?.Dispose();
        PixelShader.Dispose();
        VertexShader.Dispose();
    }
}